File: go/internal/e2e/rpc_workspace_checkpoints_e2e_test.go
package e2e

import (
	"os"
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

// Mirrors dotnet/test/E2E/RpcWorkspaceCheckpointsE2ETests.cs (snapshot category "rpc_workspace_checkpoints").
func TestRpcWorkspaceCheckpointsE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should list no checkpoints for fresh session", func(t *testing.T) {
		session := createWorkspaceRPCSession(t, client)
		defer session.Disconnect()

		result, err := session.RPC.Workspaces.ListCheckpoints(t.Context())
		if err != nil {
			t.Fatalf("Workspaces.ListCheckpoints failed: %v", err)
		}
		if result.Checkpoints == nil {
			t.Fatal("Expected non-nil Checkpoints")
		}
		if len(result.Checkpoints) != 0 {
			t.Fatalf("Expected no checkpoints for fresh session, got %+v", result.Checkpoints)
		}
	})

	t.Run("should return nil or empty content for unknown checkpoint", func(t *testing.T) {
		session := createWorkspaceRPCSession(t, client)
		defer session.Disconnect()

		result, err := session.RPC.Workspaces.ReadCheckpoint(t.Context(), &rpc.WorkspacesReadCheckpointRequest{Number: 1<<62 - 1})
		if err != nil {
			t.Fatalf("Workspaces.ReadCheckpoint failed: %v", err)
		}
		if result.Content != nil && *result.Content != "" {
			t.Fatalf("Expected nil or empty content for unknown checkpoint, got %q", *result.Content)
		}
	})

	t.Run("should return typed workspace diff result", func(t *testing.T) {
		session := createWorkspaceRPCSession(t, client)
		defer session.Disconnect()

		result, err := session.RPC.Workspaces.Diff(t.Context(), &rpc.WorkspacesDiffRequest{Mode: rpc.WorkspaceDiffModeUnstaged})
		if err != nil {
			t.Fatalf("Workspaces.Diff failed: %v", err)
		}
		if result.RequestedMode != rpc.WorkspaceDiffModeUnstaged {
			t.Fatalf("Expected RequestedMode=unstaged, got %q", result.RequestedMode)
		}
		if result.Mode != rpc.WorkspaceDiffModeUnstaged && result.Mode != rpc.WorkspaceDiffModeBranch {
			t.Fatalf("Unexpected effective diff mode %q", result.Mode)
		}
		if result.Changes == nil {
			t.Fatal("Expected non-nil Changes")
		}
		for _, change := range result.Changes {
			if strings.TrimSpace(change.Path) == "" {
				t.Fatalf("Diff change has empty path: %+v", change)
			}
			switch change.ChangeType {
			case rpc.WorkspaceDiffFileChangeTypeAdded,
				rpc.WorkspaceDiffFileChangeTypeModified,
				rpc.WorkspaceDiffFileChangeTypeDeleted,
				rpc.WorkspaceDiffFileChangeTypeRenamed:
			default:
				t.Fatalf("Unexpected diff change type %q", change.ChangeType)
			}
			_ = change.Diff
		}
	})

	t.Run("should save large paste and expose readable content", func(t *testing.T) {
		session := createWorkspaceRPCSession(t, client)
		defer session.Disconnect()
		content := strings.Repeat("Large paste payload 🚀\n", 512)

		result, err := session.RPC.Workspaces.SaveLargePaste(t.Context(), &rpc.WorkspacesSaveLargePasteRequest{Content: content})
		if err != nil {
			t.Fatalf("Workspaces.SaveLargePaste failed: %v", err)
		}
		if result.Saved == nil {
			t.Fatal("Expected SaveLargePaste to return saved descriptor")
		}
		saved := result.Saved
		if strings.TrimSpace(saved.Filename) == "" || strings.TrimSpace(saved.FilePath) == "" {
			t.Fatalf("Expected saved filename and filepath, got %+v", saved)
		}
		if saved.SizeBytes != int64(len([]byte(content))) {
			t.Fatalf("Expected SizeBytes=%d, got %d", len([]byte(content)), saved.SizeBytes)
		}

		read, readErr := session.RPC.Workspaces.ReadFile(t.Context(), &rpc.WorkspacesReadFileRequest{Path: saved.Filename})
		if readErr == nil {
			if read.Content != content {
				t.Fatalf("Expected ReadFile content to match saved paste")
			}
			return
		}

		bytes, err := os.ReadFile(saved.FilePath)
		if err != nil {
			t.Fatalf("ReadFile failed (%v), and saved file %q was not readable: %v", readErr, saved.FilePath, err)
		}
		if string(bytes) != content {
			t.Fatalf("Expected saved file content to match large paste")
		}
	})
}

func createWorkspaceRPCSession(t *testing.T, client *copilot.Client) *copilot.Session {
	t.Helper()
	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
	})
	if err != nil {
		t.Fatalf("CreateSession failed: %v", err)
	}
	return session
}

File: nodejs/test/e2e/rpc_workspace_checkpoints.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { existsSync, readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Session workspace checkpoint RPC", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    it("should list no checkpoints for fresh session", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        try {
            const result = await session.rpc.workspaces.listCheckpoints();
            expect(result.checkpoints).toEqual([]);
        } finally {
            await session.disconnect();
        }
    });

    it("should return null or empty content for unknown checkpoint", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        try {
            const result = await session.rpc.workspaces.readCheckpoint({
                number: Number.MAX_SAFE_INTEGER,
            });
            expect(result.content ?? "").toBe("");
        } finally {
            await session.disconnect();
        }
    });

    it("should return typed workspace diff result", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        try {
            const result = await session.rpc.workspaces.diff({ mode: "unstaged" });
            expect(result.requestedMode).toBe("unstaged");
            expect(["unstaged", "branch"]).toContain(result.mode);
            expect(Array.isArray(result.changes)).toBe(true);
            for (const change of result.changes) {
                expect(change.path.trim()).toBeTruthy();
                expect(["added", "modified", "deleted", "renamed"]).toContain(change.changeType);
                expect(typeof change.diff).toBe("string");
            }
        } finally {
            await session.disconnect();
        }
    });

    it("should save large paste and expose readable content", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        try {
            const content = "Large paste payload 🚀\n".repeat(512);
            const result = await session.rpc.workspaces.saveLargePaste({ content });
            const saved = result.saved;

            expect(saved).not.toBeNull();
            expect(saved!.filename.trim()).toBeTruthy();
            expect(saved!.filePath.trim()).toBeTruthy();
            expect(saved!.sizeBytes).toBe(Buffer.byteLength(content, "utf8"));

            try {
                const read = await session.rpc.workspaces.readFile({ path: saved!.filename });
                expect(read.content).toBe(content);
            } catch (err: unknown) {
                expect(existsSync(saved!.filePath)).toBe(true);
                expect(readFileSync(saved!.filePath, "utf8")).toBe(content);
                expect(err).toBeDefined();
            }
        } finally {
            await session.disconnect();
        }
    });
});
